[doc] Better document for Draft-Target-Model (DTM) speculative decoding (#3797)
NVIDIA/TensorRT-LLM: examples/draft_target_model/README.md

Ubuntu is a free and open source operating system. It is a Linux based operating system. ...
```

+ Stop triton inference server after all work is done.

```bash
pkill tritonserver
```

> + In addition, it appears better performance if both draft and target engine are deployed in one GPU (llama-7B-FP8 + llama-30B-FP8, totally 40GiB in one H100-80GiB GPU as example) than in two separate GPUs.

Reviewer: ```suggestion
+ In addition, it appears better performance can be achieved with both draft and target engines deployed on a single GPU (llama-7B-FP8 + llama-30B-FP8, for a total of 40GiB on one H100-80GiB GPU for example).
```

#### Usage of Tensor-Parallelization mode.

> + For example, we have built draft engine with TP=1 and target engine with TP=2 (both symmetrical or asymmetrical TP size are acceptable), and want to place the draft engine on GPU0, target engine on GPU1.

Reviewer: This looks odd, target engine should need 2 GPUs since TP=2?

Author: Sorry for the typo, it should be "target engine on GPU1 and GPU2".
I fixed it as (https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R242).

+ Edit model configuration the same as step `3.`, where still set `DRAFT_DEVICE_IDS="0"` and ``TARGET_DEVICE_IDS="1"` since comma is not supported in script `python3 tools/fill_template.py`.

> + Fix the device IDs in `${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt` and `${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt` manually as:

Reviewer: sed can be used for that purpose:
```
sed -i 's/\${gpu_device_ids}/0/g' ${TRITON_MODEL_REPO}/tensorrt_llm_draft/config.pbtxt
sed -i 's/\${gpu_device_ids}/1,2/g' ${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt
```

Author: Thanks for the suggestions, I add the commands to replace `gpu_device_ids` and `participant_ids` as [Line262](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R262) and [Line307](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R307). It makes the steps simple and elegant.

        string_value: "0"
    }
}

### In tensorrt_llm/config.pbtxt
parameters: {
    key: "gpu_device_ids"
    value: {
        string_value: "1,2"
    }
}
```

+ Start the triton inference server

>   + `--multi-model` must be specified if asymmetrical TP is used.

Reviewer: I thought it was required regardless of TP size?

Author: I found it worked smoothly in symmetrical TP mode without `--multi-model`, but failed in asymmetrical TP mode.
Anyway, we unify the usage as [Line269](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R269), using `--multi-model` in any TP>1 scenario.

```

+ As you see, the differences are `participant_ids` and `speculative_decoding_fast_logits`.
+ Assuming TP size of draft and target engine are `tp_draft` and `tp_target`, rank 0 is reserved for the orchestrator; rank (1 ~ tp_draft) are for draft engine; rank (tp_draft+1 ~ tp_draft+tp_target) are for target engine.
+ For example, `tp_draft`=1 and `tp_target`=2, we should fix the participant IDs in `${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt` and `${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt` manually as:

```txt
### In tensorrt_llm_draft/config.pbtxt
parameters: {
    key: "gpu_device_ids"
    value: {
        string_value: "0"
    }
}
parameters: {
    key: "participant_ids"
    value: {
        string_value: "1"
    }
}
### In tensorrt_llm/config.pbtxt
parameters: {
    key: "gpu_device_ids"
    value: {
        string_value: "1,2"
    }
}
parameters: {
    key: "participant_ids"
    value: {
        string_value: "2,3"
    }
}
```

+ Start the triton inference server.

>   + `--disable-spawn-process` to enable orchestrator mode. See [model config](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/docs/model_config.md) for more information.

Reviewer: `--multi-model` is enabling orchestrator mode. `--disable-spawn-process` is setting the pre-spawn variant of orchestrator mode.

Author: Thanks for the explanation about those parameters. I adjust related text as [Line269](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R269) and [Line316](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R316).

>   + `--world_size` must be equal to (tp_draft+tp_target+1). It is 4 in our example.

Reviewer: Might be worth documenting the +1 is for the orchestrator rank (if not documented elsewhere)

Author: We have noted it in previous version [Line237](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01L237)
But for emphasis, we move it here [Line284](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R284).

```bash
python3 scripts/launch_triton_server.py \
    --model_repo ${TRITON_MODEL_REPO} \
    --tensorrt_llm_model_name tensorrt_llm,tensorrt_llm_draft \
    --multi-model \
    --world_size 4 \
    --disable-spawn-processes
```

> + All other operations is the same as `Simple deploy` part.

Author: Fixed!

> + All other operations is the same as `Simple deploy` part.

Reviewer: L278 is fixed, but L328 still has "is the same", unless I'm misreading github diff?

> + All other operations is the same as `Simple deploy` part.

Author: No, that 's my fault of incorrect git rebase operation, now it is OK [Line328](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R328).
